### PR TITLE
various cleanups to try to reduce the use of spans inside method resolution

### PR DESCRIPTION
**Pull request by oli-obk**
definitely review commit by commit.

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/mod.rs [hunk 1]
```diff
@@ -577,9 +569,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
 
         let pick = self.probe_for_name(
-            span,
             probe::Mode::Path,
-            method_name,
+            method_name.with_span_pos(span),
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/method/mod.rs]: can you add a comment explaining why `with_span_pos` is used here? That feels weird.

Do we get weird error messages if we change this to use the old `method_name` span directly?

File: compiler/rustc_hir_typeck/src/method/mod.rs [hunk 2]
```diff
@@ -212,7 +212,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let mut candidates =
                 match self.lookup_probe(segment.ident, self_ty, call_expr, ProbeScope::AllTraits) {
                     // If we find a different result the caller probably forgot to import a trait.
-                    Ok(ref new_pick) if *new_pick != pick => {
+                    Ok(ref new_pick) if new_pick.self_ty != pick.self_ty => {
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/method/mod.rs:218]: can you not have different traits with the same `self_ty` here? I would expect us to also have to check the `Item` here

File: compiler/rustc_hir_typeck/src/method/probe.rs [hunk 3]
```diff
@@ -1343,30 +1322,37 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
...
             } else {
                 Err(MethodError::BadReturnType)
             }
         })
     }
+}
 
-    fn emit_unstable_name_collision_hint(
+impl<'tcx> Pick<'tcx> {
+    pub fn emit_unstable_name_collision_hint(
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/method/probe.rs:1335]: `fn opt_emit_unstable_name_collision_hint` or `maybe_emit_unstable_name_collision_hint`?

**Comment by lcnr** [comment]
r=me after nits

File: compiler/rustc_hir_typeck/src/method/mod.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_hir_typeck/src/method/mod.rs]: Yea... I need to dig into why [doing that](https://github.com/rust-lang/rust/pull/104394/commits/fb1a71a985c3b040e114ec1dc96a4ebe8360718f) causes a diagnostic regression

**Pull request merged**